app/Http/Controllers/AdminController.php

```diff
@@ -126,68 +125,46 @@ public function verifykeepimage (Request $request)
      */
     public function incorrect (Request $request)
     {
-        $this->reset($request->photoId);
-
-        $user = Auth::user();
-        $user->count_correctly_verified = 0;
-        $user->save();
-
-        return ['success' => true];
-    }
-
-    /**
-     * Delete litter data associated with a photo
-     */
-    protected function reset ($id)
-    {
-        $photo = Photo::find($id);
+        /** @var Photo $photo */
+        $photo = Photo::findOrFail($request->photoId);
 
         $photo->verification = 0;
         $photo->verified = 0;
         $photo->total_litter = 0;
         $photo->result_string = null;
+        $photo->save();
 
-        $categories = Photo::categories();
+        $deletedTags = $this->clearTagsAction->run($photo);
 
-        foreach ($categories as $category)
-        {
-            if ($photo->$category)
-            {
-                // hold instance of the relationship to delete
-                $d = $photo->$category;
-
-                // remove the model from the photo
-                $category_id = $category . '_id';
-                $photo->$category_id = null;
-                $photo->save();
-
-                // delete the relationship
-                $d->delete();
-            }
-        }
+        $user = User::find($photo->user_id);
+        $user->xp = max(0, $user->xp - $deletedTags['all']);
+        $user->count_correctly_verified = 0;
+        $user->save();
 
-        // persist reset verification changes
-        $photo->save();
+        $this->updateLeaderboardsAction->run($user, $photo);
+
+        return ['success' => true];
     }
 
     /**
      * Delete an image and its records
      */
     public function destroy (Request $request)
     {
-        $photo = Photo::find($request->photoId);
-        $s3 = \Storage::disk('s3');
-
-        try {
-            if (app()->environment('production'))
-            {
-                $path = substr($photo->filename, 42);
-                $s3->delete($path);
-            }
-            $photo->delete();
-        } catch (Exception $e) {
-            \Log::info(["Admin delete failed", $e]);
-        }
+        $photo = Photo::findOrFail($request->photoId);
+        $user = User::find($photo->user_id);
+
+        $this->deletePhotoAction->run($photo);
+
+        $deletedTags = $this->clearTagsAction->run($photo);
+
+        $photo->delete();
+
+        $user->xp = max(0, $user->xp - $deletedTags['all'] - 1); // Subtract 1xp for uploading
```

Comment:
Yep, I thought of it, we definitely need an Action to update user's and photo's info. But I didn't continue the refactoring further since this was supposed to touch only stuff related to photos total_litter, and I was worried we might have had conflicts from two refactoring processes happening at the same time. We can do it in another PR, or when we actually migrate it to Redis. Right now it would take some time to do it, I would like to introduce some site-wide tests for the user's xp.

Reply:
We will do this in the next PR